fiber-unix/src/scheduler.ml

```diff
@@ -446,3 +418,174 @@ let cancel_timer (timer : timer) =
 let abort t =
   (* TODO proper cleanup *)
   add_events t [ Abort ]
+
+module Process_watcher : sig
+  val init : t -> process_watcher
+
+  (** Register a new running process. *)
+  val register : process_watcher -> process -> unit
+
+  (** Send the following signal to all running processes. *)
+  val killall : process_watcher -> int -> unit
+end = struct
+  module Process_table : sig
+    val add : process_watcher -> process -> unit
+
+    val remove : process_watcher -> pid:Pid.t -> Unix.process_status -> unit
+
+    val running_count : process_watcher -> int
+
+    val iter : process_watcher -> f:(process -> unit) -> unit
+  end = struct
+    let add t job =
+      match Table.find t.table job.pid with
+      | None ->
+        Table.set t.table job.pid (Running job);
+        t.running_count <- t.running_count + 1;
+        if t.running_count = 1 then Condition.signal t.something_is_running
+      | Some (Zombie status) ->
+        Table.remove t.table job.pid;
+        add_events t.process_scheduler [ Job_completed (status, job.ivar) ]
+      | Some (Running _) -> assert false
+
+    let remove t ~pid status =
+      match Table.find t.table pid with
+      | None -> Table.set t.table pid (Zombie status)
+      | Some (Running job) ->
+        t.running_count <- t.running_count - 1;
+        Table.remove t.table pid;
+        add_events t.process_scheduler [ Job_completed (status, job.ivar) ]
+      | Some (Zombie _) -> assert false
+
+    let iter t ~f =
+      Table.iter t.table ~f:(fun data ->
+          match data with
+          | Running job -> f job
+          | Zombie _ -> ())
+
+    let running_count t = t.running_count
+  end
+
+  let register t process =
+    add_pending_events t.process_scheduler 1;
+    Mutex.lock t.mutex;
+    Process_table.add t process;
+    Mutex.unlock t.mutex
+
+  let killall t signal =
+    Mutex.lock t.mutex;
+    Process_table.iter t ~f:(fun job ->
+        try Unix.kill (Pid.to_int job.pid) signal with
+        | Unix.Unix_error _ -> ());
+    Mutex.unlock t.mutex
+
+  exception Finished of process * Unix.process_status
+
+  let wait_nonblocking_win32 t =
+    try
+      Process_table.iter t ~f:(fun job ->
+          let pid, status = Unix.waitpid [ WNOHANG ] (Pid.to_int job.pid) in
+          if pid <> 0 then raise_notrace (Finished (job, status)));
+      false
+    with
+    | Finished (job, status) ->
+      (* We need to do the [Unix.waitpid] and remove the process while holding
+         the lock, otherwise the pid might be reused in between. *)
+      Process_table.remove t ~pid:job.pid status;
+      true
+
+  let wait_win32 t =
+    while not (wait_nonblocking_win32 t) do
+      Mutex.unlock t.mutex;
+      Thread.delay 0.001;
+      Mutex.lock t.mutex
+    done
+
+  let wait_unix t =
+    Mutex.unlock t.mutex;
+    let pid, status = Unix.wait () in
+    Mutex.lock t.mutex;
```

Comment:
Shouldn't we release the mutex after the `Process.remove` call?

Reply:
Perhaps it's a bit confusing but this is how it's done in dune. The next call is always `Condition.wait` which will unlock the mutex.